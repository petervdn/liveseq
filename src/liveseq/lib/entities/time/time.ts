import type { Opaque } from 'type-fest';

export type Beats = Opaque<number, 'Beats'>;
export type Bpm = Opaque<number, 'Bpm'>;
export type TimeInSeconds = Opaque<number, 'TimeInSeconds'>;
