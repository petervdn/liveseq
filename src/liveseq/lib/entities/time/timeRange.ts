import type { Beats, TimeInSeconds } from './time';

export type BeatsRange = {
  start: Beats;
  end: Beats;
};

export type TimeRange = {
  start: TimeInSeconds;
  end: TimeInSeconds;
};
