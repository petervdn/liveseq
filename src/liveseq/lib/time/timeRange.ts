import type { TimeInSeconds } from '../types';

export type TimeRange = {
  start: TimeInSeconds;
  end: TimeInSeconds;
};
