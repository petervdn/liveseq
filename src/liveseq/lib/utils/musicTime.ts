export type MusicTime = [
  bars: number,
  beats?: number,
  sixteenths?: number,
  remainingSixteenths?: number,
];

// todo: these should be part of a musicTime
type MusicTimeOptions = {
  sixteenthsPerBeat?: number;
  beatsPerBar?: number;
};

const DEFAULT_BEATS_PER_BAR = 4;
const DEFAULT_SIXTEENTHS_PER_BEAT = 4;

export const musicTimeToBeats = (
  time: MusicTime,
  {
    beatsPerBar = DEFAULT_BEATS_PER_BAR,
    sixteenthsPerBeat = DEFAULT_SIXTEENTHS_PER_BEAT,
  }: MusicTimeOptions = {},
): number => {
  const [bars, beats = 0, sixteenths = 0, remainingSixteenth = 0] = time;

  return bars * beatsPerBar + beats + (sixteenths + remainingSixteenth) / sixteenthsPerBeat;
};

export const normalizeMusicTime = (
  time: MusicTime,
  {
    beatsPerBar = DEFAULT_BEATS_PER_BAR,
    sixteenthsPerBeat = DEFAULT_SIXTEENTHS_PER_BEAT,
  }: MusicTimeOptions = {},
): MusicTime => {
  const totalBeats = musicTimeToBeats(time, { sixteenthsPerBeat, beatsPerBar });

  const totalSixteenths = totalBeats * sixteenthsPerBeat;
  const flooredSixteenths = Math.floor(totalSixteenths);
  const sixteenthsPerBar = sixteenthsPerBeat * beatsPerBar;
  const bars = Math.floor(flooredSixteenths / sixteenthsPerBar);
  const beats = Math.floor((flooredSixteenths - bars * sixteenthsPerBar) / sixteenthsPerBeat);

  return [
    bars,
    beats,
    flooredSixteenths - bars * sixteenthsPerBar - beats * sixteenthsPerBeat,
    totalSixteenths - flooredSixteenths,
  ];
};

export const beatsToTime = (beats: number, bpm: number): number => (beats * 60) / bpm;

export const timeToBeats = (time: number, bpm: number): number => (time / 60) * bpm;

export const musicTimeToTime = (
  musicTime: MusicTime,
  bpm: number,
  {
    beatsPerBar = DEFAULT_BEATS_PER_BAR,
    sixteenthsPerBeat = DEFAULT_SIXTEENTHS_PER_BEAT,
  }: MusicTimeOptions = {},
): number => beatsToTime(musicTimeToBeats(musicTime, { beatsPerBar, sixteenthsPerBeat }), bpm);

export const timeToMusicTime = (time: number, bpm: number): MusicTime => {
  return [0, timeToBeats(time, bpm)];
};

// is a before b
export const isBefore = (a: MusicTime, b: MusicTime) => musicTimeToBeats(a) < musicTimeToBeats(b);

export const isSameTime = (a: MusicTime, b: MusicTime) =>
  musicTimeToBeats(a) === musicTimeToBeats(b);

export const addMusicTime = (a: MusicTime, b: MusicTime): MusicTime => {
  const [a1, a2 = 0, a3 = 0, a4 = 0] = a;
  const [b1, b2 = 0, b3 = 0, b4 = 0] = b;

  return normalizeMusicTime([a1 + b1, a2 + b2, a3 + b3, a4 + b4]);
};

// is rangeA intersecting rangeB
export const isInRange = (
  aStart: MusicTime,
  aEnd: MusicTime,
  bStart: MusicTime,
  bEnd: MusicTime,
) => {
  return !(isBefore(aEnd, bStart) || isBefore(bEnd, aStart));
};
