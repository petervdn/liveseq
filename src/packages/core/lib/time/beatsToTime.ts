import type { Beats, Bpm, TimeInSeconds } from '../types';

export const beatsToTime = (beats: Beats, bpm: Bpm): TimeInSeconds => {
  return ((beats * 60) / bpm) as TimeInSeconds;
};
