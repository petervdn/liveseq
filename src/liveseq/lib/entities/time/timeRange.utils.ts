import { timeToBeats } from './musicTime';
import type { BeatsRange, TimeRange } from './timeRange';
import type { Beats, Bpm } from './time';

// TODO: consolidate this stuff with musicTime stuff in their own "timing" directory

export const createRangeFromDuration = (duration: Beats, start = 0 as Beats) => {
  const end = (start + duration) as Beats;

  return {
    start,
    end,
  };
};

export const timeRangeToBeatsRange = (timeRange: TimeRange, bpm: Bpm): BeatsRange => {
  return {
    start: timeToBeats(timeRange.start, bpm),
    end: timeToBeats(timeRange.end, bpm),
  };
};

export const getRangeDuration = (range: BeatsRange): Beats => {
  return (range.end - range.start) as Beats;
};

export const splitRange = (range: BeatsRange, splitAt: Beats): Array<BeatsRange> => {
  const isBetween = range.start < splitAt && range.end > splitAt;

  return isBetween
    ? [
        {
          start: range.start,
          end: splitAt,
        },
        {
          start: splitAt,
          end: range.end,
        },
      ]
    : [range];
};

export const setStart = (range: BeatsRange, start: Beats): BeatsRange => {
  const newStart = Math.min(start, range.end) as Beats;
  return {
    start: newStart,
    end: range.end,
  };
};

export const setEnd = (range: BeatsRange, end: Beats): BeatsRange => {
  const newEnd = Math.max(end, range.start) as Beats;
  return {
    start: range.start,
    end: newEnd,
  };
};

export const splitRangeByDuration = (range: BeatsRange, duration: Beats): Array<BeatsRange> => {
  return splitRange(range, (range.start + duration) as Beats);
};

// moves whole range to new start
export const moveRange = (range: BeatsRange, newStart: Beats): BeatsRange => {
  const difference = newStart - range.start;

  return {
    start: range.start + difference,
    end: range.end + difference,
  } as BeatsRange;
};

// is rangeA intersecting rangeB
export const isInRange = (rangeA: BeatsRange, rangeB: BeatsRange) => {
  // todo: btw this implies that starts are before the end, which might not be the case
  return !(rangeA.end < rangeB.start || rangeB.end < rangeA.start);
};

// given a range and a number of loops, what is the resulting range?
const getLoopedRange = (range: BeatsRange, loops: number): BeatsRange => {
  const times = Math.max(0, loops + 1);
  const newDuration = (getRangeDuration(range) * times) as Beats;
  return createRangeFromDuration(newDuration, range.start);
};

const clampRange = (range: BeatsRange, rangeLimit: BeatsRange): BeatsRange => {
  return {
    start: Math.max(range.start, rangeLimit.start),
    end: Math.min(range.end, rangeLimit.end),
  } as BeatsRange;
};

// TODO: this could be optimized
// returns an array of wrapped ranges representing the range in each loop iteration
export const getWrappedRanges = (
  rangeToWrap: BeatsRange,
  rangeLimit: BeatsRange,
  loops = 0,
): Array<BeatsRange> => {
  const loopedLimit = getLoopedRange(rangeLimit, loops);
  const clampedRangeToWrap = clampRange(rangeToWrap, loopedLimit);

  const startDifference = clampedRangeToWrap.start - rangeLimit.start;

  const clampedRangeDuration = getRangeDuration(clampedRangeToWrap);
  const rangeLimitDuration = getRangeDuration(rangeLimit);

  // if rangeLimit and clampedRangeToWrap start at the same time and have same duration, coveredArea will be 1
  const coveredArea = (startDifference + clampedRangeDuration) / rangeLimitDuration;

  // how many ranges will need to be generated to cover the wrapped range
  const iterations = Math.ceil(coveredArea);

  if (iterations < 1) return [];

  // the result will be an array filled with rangeLimit
  // but the first item and last item are special cases
  const result = new Array(iterations).fill(rangeLimit);
  // gotta set the start time of the first item...
  result[0] = setStart(result[0], clampedRangeToWrap.start);

  if (result.length === 1) return result;

  // ...and end time of the last item
  const durationFraction = coveredArea % 1 || 1;
  result[result.length - 1] = createRangeFromDuration(
    (rangeLimitDuration * durationFraction) as Beats,
    rangeLimit.start,
  );

  return result;
};

export const getItemsInRange = <T extends BeatsRange>(range: BeatsRange, items: Array<T>) => {
  return items.filter((item) => {
    return isInRange(item, range);
  });
};

export const addToRange = <T extends BeatsRange>(range: T, offsetBy: Beats): T => {
  return {
    ...range,
    start: range.start + offsetBy,
    end: range.end + offsetBy,
  };
};

export const subtractFromRange = <T extends BeatsRange>(range: T, offsetBy: Beats): T => {
  return {
    ...range,
    start: range.start - offsetBy,
    end: range.end - offsetBy,
  };
};
