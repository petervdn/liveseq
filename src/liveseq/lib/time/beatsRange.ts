import { timeToBeats } from './musicTime';
import type { TimeRange } from './timeRange';
import type { Beats, Bpm } from './time';
import { wrapInRange } from '../utils/wrapInRange';

export type BeatsRange = {
  start: Beats;
  end: Beats;
};

export const createRange = (start: Beats, end: Beats): BeatsRange => {
  return {
    start,
    end,
  };
};

export const createRangeFromDuration = (duration: Beats, start = 0 as Beats) => {
  const end = (start + duration) as Beats;

  return {
    start,
    end,
  };
};

export const timeRangeToBeatsRange = (timeRange: TimeRange, bpm: Bpm): BeatsRange => {
  return {
    start: timeToBeats(timeRange.start, bpm),
    end: timeToBeats(timeRange.end, bpm),
  };
};

export const getRangeDuration = (range: BeatsRange): Beats => {
  return (range.end - range.start) as Beats;
};

export const splitRange = (range: BeatsRange, splitAt: Beats): Array<BeatsRange> => {
  const isBetween = range.start < splitAt && range.end > splitAt;

  return isBetween
    ? [
        {
          start: range.start,
          end: splitAt,
        },
        {
          start: splitAt,
          end: range.end,
        },
      ]
    : [range];
};

export const setStart = <T extends BeatsRange>(range: T, start: Beats): T => {
  return {
    ...range,
    start,
    end: Math.max(range.end, start) as Beats,
  };
};

export const setEnd = <T extends BeatsRange>(range: T, end: Beats): T => {
  return {
    ...range,
    start: Math.min(end, range.start) as Beats,
    end,
  };
};

export const splitRangeByDuration = (range: BeatsRange, duration: Beats): Array<BeatsRange> => {
  return splitRange(range, (range.start + duration) as Beats);
};

// moves whole range to new start
export const moveRange = (range: BeatsRange, newStart: Beats): BeatsRange => {
  const difference = newStart - range.start;

  return {
    start: range.start + difference,
    end: range.end + difference,
  } as BeatsRange;
};

// is rangeA intersecting rangeB
export const isInRange = (rangeA: BeatsRange, rangeB: BeatsRange) => {
  // todo: btw this implies that starts are before the end, which might not be the case
  return !(rangeA.end <= rangeB.start || rangeA.start >= rangeB.end);
};

// given a range and a number of loops, what is the resulting range?
export const getLoopedRange = (range: BeatsRange, loops: number): BeatsRange => {
  const times = Math.max(0, loops + 1);
  const newDuration = (getRangeDuration(range) * times) as Beats;
  return createRangeFromDuration(newDuration, range.start);
};

// might clamp to an invalid duration, so might return null
export const clampRange = (range: BeatsRange, rangeLimit: BeatsRange): BeatsRange | null => {
  const newRange = {
    start: Math.max(range.start, rangeLimit.start),
    end: Math.min(range.end, rangeLimit.end),
  } as BeatsRange;

  const duration = getRangeDuration(newRange);
  return duration > 0 ? newRange : null;
};

// TODO: this could be optimized
// returns an array of wrapped ranges representing the range in each loop iteration
export const getWrappedRanges = (
  rangeToWrap: BeatsRange,
  rangeLimit: BeatsRange,
  limitLoops = 0,
): Array<BeatsRange & { offset: Beats }> => {
  const times = Math.max(0, limitLoops + 1);
  const rangeLimitDuration = getRangeDuration(rangeLimit);
  const loopedLimitDuration = (rangeLimitDuration * times) as Beats;
  const loopedLimit = createRangeFromDuration(loopedLimitDuration, rangeLimit.start);
  const clampedRangeToWrap = clampRange(rangeToWrap, loopedLimit);

  if (!clampedRangeToWrap) return [];

  const clampedRangeDuration = getRangeDuration(clampedRangeToWrap);

  const startDifference = clampedRangeToWrap.start - rangeLimit.start;
  const firstIteration = Math.floor(startDifference / rangeLimitDuration); // todo: might need to ceil or something
  const wrappedStartDifference = wrapInRange(startDifference, rangeLimit);
  // if rangeLimit and clampedRangeToWrap start at the same time and have same duration, coveredArea will be 1
  const coveredArea = (wrappedStartDifference + clampedRangeDuration) / rangeLimitDuration;

  // how many ranges will need to be generated to cover the wrapped range
  const totalRanges = Math.ceil(coveredArea);

  // console.log({
  //   startDifference,
  //   rangeToWrap,
  //   rangeLimit,
  //   clampedRangeToWrap,
  //   loopedLimit,
  //   firstIteration,
  //   coveredArea,
  //   totalRanges,
  //   // clampedRangeDuration,
  //   // rangeLimitDuration,
  // });

  // the result will be an array filled with rangeLimit
  // but the first item and last item are special cases
  // eslint-disable-next-line @typescript-eslint/naming-convention
  const result = Array.from({ length: totalRanges }, (_, index) => {
    return {
      ...rangeLimit,
      offset: ((index + firstIteration) * rangeLimitDuration) as Beats,
    };
  });

  // TODO: this was wrong, but must be included
  // gotta set the start time of the first item...
  // mutation!
  result[0] = setStart(result[0], wrappedStartDifference as Beats);

  if (result.length === 1) return result;

  // ...and end time of the last item
  const durationFraction = coveredArea % 1 || 1;
  const lastItem = result[result.length - 1];
  // mutation!
  result[result.length - 1] = {
    ...lastItem,
    ...createRangeFromDuration((rangeLimitDuration * durationFraction) as Beats, rangeLimit.start),
  };

  return result;
};

export const getItemsInRange = <T extends BeatsRange>(range: BeatsRange, items: Array<T>) => {
  return items.filter((item) => {
    return isInRange(item, range);
  });
};

export const addToRange = <T extends BeatsRange>(range: T, offsetBy: Beats): T => {
  return {
    ...range,
    start: range.start + offsetBy,
    end: range.end + offsetBy,
  };
};

export const subtractFromRange = <T extends BeatsRange>(range: T, offsetBy: Beats): T => {
  return {
    ...range,
    start: range.start - offsetBy,
    end: range.end - offsetBy,
  };
};
