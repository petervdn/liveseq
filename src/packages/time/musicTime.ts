import type { Beats, Bpm, TimeInSeconds } from '../core/lib/types';
import { beatsToTime } from '../core/lib/time/beatsToTime';
import { timeToBeats } from '../core/lib/time/timeToBeats';

export type MusicTime = [
  bars: number,
  beats?: number,
  sixteenths?: number,
  remainingSixteenths?: number,
];

// todo: these should be part of a musicTime
type MusicTimeOptions = {
  sixteenthsPerBeat?: number;
  beatsPerBar?: number;
};

const DEFAULT_BEATS_PER_BAR = 4;
const DEFAULT_SIXTEENTHS_PER_BEAT = 4;

export const musicTimeToBeats = (
  time: MusicTime,
  {
    beatsPerBar = DEFAULT_BEATS_PER_BAR,
    sixteenthsPerBeat = DEFAULT_SIXTEENTHS_PER_BEAT,
  }: MusicTimeOptions = {},
): Beats => {
  const [bars, beats = 0, sixteenths = 0, remainingSixteenth = 0] = time;

  return (bars * beatsPerBar +
    beats +
    (sixteenths + remainingSixteenth) / sixteenthsPerBeat) as Beats;
};

export const normalizeMusicTime = (
  time: MusicTime,
  {
    beatsPerBar = DEFAULT_BEATS_PER_BAR,
    sixteenthsPerBeat = DEFAULT_SIXTEENTHS_PER_BEAT,
  }: MusicTimeOptions = {},
): MusicTime => {
  const end = musicTimeToBeats(time, { sixteenthsPerBeat, beatsPerBar });

  const totalSixteenths = end * sixteenthsPerBeat;
  const flooredSixteenths = Math.floor(totalSixteenths);
  const sixteenthsPerBar = sixteenthsPerBeat * beatsPerBar;
  const bars = Math.floor(flooredSixteenths / sixteenthsPerBar);
  const beats = Math.floor((flooredSixteenths - bars * sixteenthsPerBar) / sixteenthsPerBeat);

  return [
    bars,
    beats,
    flooredSixteenths - bars * sixteenthsPerBar - beats * sixteenthsPerBeat,
    totalSixteenths - flooredSixteenths,
  ];
};

export const musicTimeToTime = (
  musicTime: MusicTime,
  bpm: Bpm,
  {
    beatsPerBar = DEFAULT_BEATS_PER_BAR,
    sixteenthsPerBeat = DEFAULT_SIXTEENTHS_PER_BEAT,
  }: MusicTimeOptions = {},
): number => beatsToTime(musicTimeToBeats(musicTime, { beatsPerBar, sixteenthsPerBeat }), bpm);

export const timeToMusicTime = (time: TimeInSeconds, bpm: Bpm): MusicTime =>
  normalizeMusicTime([0, timeToBeats(time, bpm)]);

// is a before b
export const isBefore = (a: MusicTime, b: MusicTime) => musicTimeToBeats(a) < musicTimeToBeats(b);

export const isSameTime = (a: MusicTime, b: MusicTime) => {
  return musicTimeToBeats(a) === musicTimeToBeats(b);
};

export const addMusicTime = (a: MusicTime, b: MusicTime) => {
  return normalizeMusicTime([0, musicTimeToBeats(a) + musicTimeToBeats(b)]);
};
