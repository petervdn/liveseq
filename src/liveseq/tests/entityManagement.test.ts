import { Beats, createLiveseq, Liveseq } from '..';
import { musicTimeToBeats } from '../lib/time/musicTime';

const addAllTypesOfEntities = (liveseq: Liveseq) => {
  const clipId = liveseq.addClip({
    type: 'noteClip',
    duration: 10 as Beats,
    notes: [],
  });

  const timelineId = liveseq.addTimeline({
    name: 'Timeline Name',
    duration: musicTimeToBeats([1, 0, 0]),
    clips: [
      {
        clipId,
        start: musicTimeToBeats([0, 0, 0]),
        end: musicTimeToBeats([1, 0, 0]),
      },
    ],
  });

  const slotId = liveseq.addSlot({
    type: 'timelineSlot',
    timelineId,
    loops: 0,
  });

  liveseq.addScene({
    eventActions: {
      enter: [
        {
          type: 'playSlots',
          slotIds: [slotId],
        },
      ],
    },
  });

  liveseq.addChannel({
    type: 'instrumentChannel',
    instrumentId: liveseq.addInstrument({
      type: 'samplerInstrument',
    }),
    slotIds: [slotId],
  });
};

it('adds all types of entities correctly', () => {
  const liveseq = createLiveseq();

  addAllTypesOfEntities(liveseq);

  expect(liveseq.getProject().entities).toMatchSnapshot();

  addAllTypesOfEntities(liveseq);

  expect(liveseq.getProject().entities).toMatchSnapshot();
});
