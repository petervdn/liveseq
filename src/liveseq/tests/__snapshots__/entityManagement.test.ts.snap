// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds all types of entities correctly 1`] = `
Object {
  "channels": Array [
    Object {
      "id": "channel_0",
      "instrumentId": "instrument_0",
      "slotIds": Array [
        "slot_0",
      ],
      "type": "instrumentChannel",
    },
  ],
  "clips": Array [
    Object {
      "duration": 10,
      "id": "clip_0",
      "notes": Array [],
      "type": "noteClip",
    },
  ],
  "instruments": Array [
    Object {
      "id": "instrument_0",
      "type": "samplerInstrument",
    },
  ],
  "samples": Array [],
  "scenes": Array [
    Object {
      "eventActions": Object {
        "enter": Array [
          Object {
            "slotIds": Array [
              "slot_0",
            ],
            "type": "playSlots",
          },
        ],
      },
      "id": "scene_0",
    },
  ],
  "slots": Array [
    Object {
      "id": "slot_0",
      "loops": 0,
      "timelineId": "timeline_0",
      "type": "timelineSlot",
    },
  ],
  "timelines": Array [
    Object {
      "clips": Array [
        Object {
          "clipId": "clip_0",
          "end": 4,
          "start": 0,
        },
      ],
      "duration": 4,
      "id": "timeline_0",
      "name": "Timeline Name",
    },
  ],
}
`;

exports[`adds all types of entities correctly 2`] = `
Object {
  "channels": Array [
    Object {
      "id": "channel_0",
      "instrumentId": "instrument_0",
      "slotIds": Array [
        "slot_0",
      ],
      "type": "instrumentChannel",
    },
    Object {
      "id": "channel_1",
      "instrumentId": "instrument_1",
      "slotIds": Array [
        "slot_1",
      ],
      "type": "instrumentChannel",
    },
  ],
  "clips": Array [
    Object {
      "duration": 10,
      "id": "clip_0",
      "notes": Array [],
      "type": "noteClip",
    },
    Object {
      "duration": 10,
      "id": "clip_1",
      "notes": Array [],
      "type": "noteClip",
    },
  ],
  "instruments": Array [
    Object {
      "id": "instrument_0",
      "type": "samplerInstrument",
    },
    Object {
      "id": "instrument_1",
      "type": "samplerInstrument",
    },
  ],
  "samples": Array [],
  "scenes": Array [
    Object {
      "eventActions": Object {
        "enter": Array [
          Object {
            "slotIds": Array [
              "slot_0",
            ],
            "type": "playSlots",
          },
        ],
      },
      "id": "scene_0",
    },
    Object {
      "eventActions": Object {
        "enter": Array [
          Object {
            "slotIds": Array [
              "slot_1",
            ],
            "type": "playSlots",
          },
        ],
      },
      "id": "scene_1",
    },
  ],
  "slots": Array [
    Object {
      "id": "slot_0",
      "loops": 0,
      "timelineId": "timeline_0",
      "type": "timelineSlot",
    },
    Object {
      "id": "slot_1",
      "loops": 0,
      "timelineId": "timeline_1",
      "type": "timelineSlot",
    },
  ],
  "timelines": Array [
    Object {
      "clips": Array [
        Object {
          "clipId": "clip_0",
          "end": 4,
          "start": 0,
        },
      ],
      "duration": 4,
      "id": "timeline_0",
      "name": "Timeline Name",
    },
    Object {
      "clips": Array [
        Object {
          "clipId": "clip_1",
          "end": 4,
          "start": 0,
        },
      ],
      "duration": 4,
      "id": "timeline_1",
      "name": "Timeline Name",
    },
  ],
}
`;
